app/src/composables/use-navigation-guard.ts

```diff
@@ -0,0 +1,22 @@
+import { type Ref, onBeforeMount, onBeforeUnmount } from 'vue';
+import { onBeforeRouteUpdate, onBeforeRouteLeave, NavigationGuard } from 'vue-router';
+
+export function useNavigationGuard(locked: Ref<boolean>, guard: NavigationGuard) {
+	onBeforeRouteUpdate(guard);
+	onBeforeRouteLeave(guard);
+
+	const beforeUnload = (event: BeforeUnloadEvent) => {
+		if (locked.value) {
+			event.preventDefault();
+			event.returnValue = '';
+		}
+	};
+
+	const addListener = () => window.addEventListener('beforeunload', beforeUnload);
+	const removeListener = () => window.removeEventListener('beforeunload', beforeUnload);
+
+	onBeforeMount(addListener);
+	onBeforeUnmount(removeListener);
+
+	return { removeGuard: removeListener };
+}
```

Comment:
Let's
- simplify with [`useEventListener`](https://vueuse.org/core/useEventListener/) from VueUse
- no longer set `event.returnValue`, it's [deprecated](https://developer.mozilla.org/en-US/docs/Web/API/Event/returnValue)
- not return `removeGuard`, see comment in `extensions.vue`

```suggestion
import { useEventListener } from '@vueuse/core';
import { type Ref } from 'vue';
import { NavigationGuard, onBeforeRouteLeave, onBeforeRouteUpdate } from 'vue-router';

export function useNavigationGuard(locked: Ref<boolean>, guard: NavigationGuard) {
	onBeforeRouteUpdate(guard);
	onBeforeRouteLeave(guard);

	useEventListener(document, 'beforeunload', (event) => {
		if (locked.value) {
			event.preventDefault();
		}
	});
}
```